### ISSUE TITLE
"--backend vs" Does not produce functioning MSVC project definitions involving local subproject libraries

### ISSUE THREAD

**Issue by btgoodwin** [mesonbuild/meson#11527]
**Describe the bug**
The Windows environment requires all libraries to be visible from the `PATH` environment variable when an executable is run.  When using  `--backend vs`, the generated project and solution files are lacking appropriate extensions to the `PATH` to ensure runtime linking works properly (when trying to run a test from the generated Visual Studio solution).  Conversely, when running `meson test ...` from the command line, the `PATH` is extended for the location of all declared dependencies (libraries) so that the test can execute within that environment.

While it is possible to edit the project's debugger settings each time one configures the meson project, that is a brittle and tedious work-around to what appears to be a missing feature in the project and solution generation module.  In my particular case, the main reason for using this feature is to utilize the Visual Studio debugger since meson supports only gdb (it appears) and attempting to get VSCode's `'type': 'cppvsdbg'` to work with Meson's internally-held environment functionality have come up dry as well (like trying to use a pre-launch task that runs `meson devenv -C builddir` blocks the debugger process until the terminal exits).

**To Reproduce**

I've created a minimal project repository with instructions [here](https://github.com/btgoodwin/meson-example-project/blob/main/README.md).

**Expected behavior**

I would expect the use of `--backend vs` to produce a solution file with functioning projects without having to manually edit the files each time I modify and then have to reconfigure the meson build environment.

**system parameters**
* Native Build
* Windows 11 21H2
* Python 3.9.5
* Meson 1.0.1
* Ninja 1.11.0

**Comment by eli-schwartz** [comment]
I'm not involved with the vs backend, but I do know that `meson devenv` is, broadly speaking, intended to make stuff like this work -- is it possible to run VS *from* a devenv shell? Maybe awkward too...

...

I'm assuming there's some property that can be set on a vs solution target in the XML, which defines the PATH used for running the executable via the Visual Studio GUI? Can you point me to documentation for this? Alternatively maybe @jpakkane knows.

It should be feasible to just add this to the target generation, we do have all the information needed at that time.

**Comment by btgoodwin** [comment]
That's correct, it's in the XML for the referenced project (i.e., the solution points to the `.vcxproj` file).  When I edit the _Environment_ to update the `PATH` variable in the debugger settings from within Visual Studio to include:

```
PATH=$(ProjectDir)subprojects\libexample;$(PATH)
```

Visual Studio creates a `<project>.vcxproj.user` file with the following contents, which I believe are merged with the original's settings by VS:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='debug|x64'">
    <LocalDebuggerEnvironment>PATH=$(ProjectDir)subprojects\libexample;$(PATH)
$(LocalDebuggerEnvironment)</LocalDebuggerEnvironment>
    <DebuggerFlavor>WindowsLocalDebugger</DebuggerFlavor>
  </PropertyGroup>
</Project>
```

The end result is I can now run the `main` target and debug it without getting the library error at runtime.  However each time I reconfigure the project, I have to re-do that setting.

**Comment by btgoodwin** [comment]
Also, you're correct, `devenv` would help here if I could use it as a pre-launch task.  But since it essentially blocks until exited, I don't really have a way to use it in that capacity.  I tried defining an integrated terminal via VSCode's `settings.json` to run `meson devenv -C builddir` to be the default, however it does not seem to apply to how the `launch.json` brings up the terminal for the debugger (the launch config has `"console": "integratedTerminal"` set, which should cause it to use that defined default terminal but again it creates an integrated terminal with the meson devenv running and then runs the debugger somewhere else).

I did just find a work-around here for getting a debugger up in VSCode that utilizes the result of `devenv`, however.  It's not a fix for this though; it only gets me to a point where I don't have to manually maintain the project file(s).

1. Create a batch file in the root, like `devenv-dump.bat`, with the contents:
    ```batch
    meson devenv --dump -C %1 > %1\devenv.env 2>&1
    ```
2. Create a VSCode `task.json` definition like this:
    ```json
    {
      "version": "2.0.0",
      "tasks": [
        {
          "type": "process",
          "label": "Meson Dump Devenv",
          "command": "echo Only needed for Windows", // Default, do nothing
          "options": {"cwd": "${workspaceFolder}"},
          "windows": {
              "command": "devenv-dump.bat",
              "args": ["${config:mesonbuild.buildFolder}"],
          }
        }
      ]
    }
    ```
3. Define in the `.vscode/launch.json` a debug configuration for `cl`:
    ```json
    {
      "version": "0.2.0",
      "configurations": [
        {
          "name": "cl.exe - Debug active file",
          "type":"cppvsdbg",
          "request": "launch",
          "program": "${workspaceFolder}\\${config:mesonbuild.buildFolder}\\${relativeFileDirname}\\${fileBasenameNoExtension}.exe",
          "stopAtEntry": true,
          "cwd": "${workspaceFolder}",
          "console": "integratedTerminal",
          "preLaunchTask": "Meson Dump Devenv",
          "envFile": "${workspaceFolder}\\${config:mesonbuild.buildFolder}\\devenv.env"
        }
      ]
    }
    ```
As long as I'm picking a source file that results in an executable target, and I've already run `meson compile`, I can now use the `Debug: Start Debugging` CTRL+SHIFT+P task without any issue I can see (so far...).

**Comment by btgoodwin** [comment]
> I'm not involved with the vs backend, but I do know that `meson devenv` is, broadly speaking, intended to make stuff like this work -- is it possible to run VS _from_ a devenv shell? Maybe awkward too...

I hadn't thought of that...but that does work:

```
meson setup builddir --backend vs
meson devenv -C builddir
# shell instance updated with paths, etc.
devenv "builddir/<name of solution>.sln"
```

You're right thought it's a little awkward.  If I update the meson build environment at all, I would need to exit VS, exit the devenv, and re-run those steps.

**Comment by eli-schwartz** [comment]
Yup, `devenv` by default drops you into a PowerShell or cmd prompt. You can also use it as a wrapper for running the program with `meson devenv myprog.exe` but then you need to prepend it to every command, and then it runs a relatively hefty python program first, every single time...

I can take a look at trying to add a LocalDebuggerEnvironment key if no one beats me to it, so that this doesn't require hacks.

**Comment by btgoodwin** [comment]
Thank you!

FWIW my idea of piping the devenv output to a `.env` file, and then ingesting that only worked a few times.  Now meson is writing out `Activating VS 17.4.2`, which gets redirected into the file as well, making it impossible to parse.  I don't know what I changed in my VSCode environment that caused this to happen; I haven't re-opened the session or anything.  Several debug runs it worked, the next it didn't, and it has been broken ever since. :-/

**Comment by eli-schwartz** [comment]
Really, the solution is just to avoid redirecting stdout but using #11295 instead, I think.

That line is probably coming from somewhere in which meson is executing vcvars*.bat on your behalf, probably as a prerequisite to defining the devenv (?).

**Comment by xclaesse** [comment]
> Now meson is writing out `Activating VS 17.4.2`

oh, that's unexpected side effect of https://github.com/mesonbuild/meson/pull/11277, we should silence it when using --dump.

**Comment by xclaesse** [comment]
@btgoodwin --dump has not been tested much on Windows. I suspect we would need to add powershell and cmd to `--dump-format`, to format like `${env:FOO}` and `%FOO`.
